### ISSUE TITLE
[stable/prometheus-operator] Prometheus CRD storage spec metadata ignored in upgrades, creates new PVC

### ISSUE THREAD

**Issue by gracedo** [helm/charts#21290]
**Describe the bug**
On upgrade from version 8.3.3 to 8.7.0, we discovered that a new PVC was created instead of using the PVC prior to upgrade. Looking at the Prometheus resource, we noticed that the `prometheus.prometheusSpec.storageSpec.volumeClaimTemplate.metadata` field was set to empty, ignoring our configs (we set `prometheus.prometheusSpec.storageSpec.volumeClaimTemplate.metadata` to have a `name` field).
e.g.
```
kubectl get prometheus -oyaml
...
    storage:
      volumeClaimTemplate:
        metadata: {}
        spec:
          accessModes:
          - ReadWriteOnce
          resources:
            requests:
              storage: 50Gi
...
```

It looks like maybe we can tie this bug back to https://github.com/helm/charts/pull/20419, which introduced a change in the CRD to use the standard object's metadata.

**Version of Helm and Kubernetes**:
Helm v2.16.1
Kubernetes v1.16.4

**Which chart**:
stable/prometheus-operator

**What happened**:
New PVC was created (different name) and used by Prometheus

**What you expected to happen**:
Expected previously-created PVC to be used. This resulted in Prometheus using a new PV which meant old data wasn't getting shown.

**How to reproduce it** (as minimally and precisely as possible):
Upgrade the chart from 8.3.3 to 8.7.0 passing in the following (example) values to set the metadata name field
```
prometheus:
  prometheusSpec:
    storageSpec:
      volumeClaimTemplate:
        metadata:
          name: test
        spec:
          accessModes: ["ReadWriteOnce"]
          resources:
            requests:
              storage: 50Gi
```

**Anything else we need to know**:

**Comment by gracedo** [comment]
@vsliouniaev would you mind taking a look at this bug when convenient? Thank you!

**Comment by vsliouniaev** [comment]
It's quite possible that this is related. There have been two new versions of prometheus-operator released since, would you mind checking those?

**Comment by gracedo** [comment]
Upgrading from v8.3.3 to v8.12.1 (latest) did not fix the bug.

I also did some more testing to try to narrow down versions:
- Installed at v8.0.0: `prometheus.prometheusSpec.storageSpec.volumeClaimTemplate.metadata` fields correctly set.
- I then ran upgrades bumping each minor version at a time until `prometheus.prometheusSpec.storageSpec.volumeClaimTemplate.metadata` was incorrectly set to `{}`, which was once I upgraded from v8.5.0 to v8.6.0.
- I then ran downgrades bumping down minor versions and tried all the way down to v7.0.0 and stopped there - but never saw the metadata get populated.
- Installed v8.5.15: `prometheus.prometheusSpec.storageSpec.volumeClaimTemplate.metadata` fields correctly set.
- Installed v8.6.0: : `prometheus.prometheusSpec.storageSpec.volumeClaimTemplate.metadata` fields empty.

So it does like this was introduced between v8.5.15 and v8.6.0, which matches the CRD change mentioned in the PR description https://github.com/helm/charts/pull/20419

**Comment by gracedo** [comment]
@vsliouniaev @bismarck @gianrubio Would you mind taking a look into this PVC issue? Let me know if there's anything I can help with!

**Comment by mdbraber** [comment]
Same bug here - @gracedo have you found a fix?

**Comment by gracedo** [comment]
@mdbraber The fix in coreos/prometheus-operator was done here: https://github.com/coreos/prometheus-operator/pull/3115 and released in [v0.39.0](https://github.com/coreos/prometheus-operator/releases/tag/v0.39.0). There was a helm chart PR https://github.com/helm/charts/pull/22338 to upgrade the operator to v0.39.0 but [due to this version requiring kubernetes 1.16+ and unclear roadmap/future for the chart](https://github.com/helm/charts/pull/22338#issuecomment-628524979), it looks like it's been put on pause for the time being.

**Comment by mdbraber** [comment]
@gracedo thanks! Even when using an already updated chart (like from thiss repository: https://github.com/iwilltry42/charts) I'm not able to get it to work unfortunately with v0.39.0. I've detailed it here: https://github.com/coreos/prometheus-operator/issues/3093#issuecomment-636716487
